Added Fuzz API for addOn Fuzz

davy320 opened the pull request:
This pull request contains the FuzzAPI v1.0 for addOn fuzz, all the help documentation required has been added, please check Github issue: 
[Github Issue](https://github.com/zaproxy/zaproxy/issues/1689): https://github.com/zaproxy/zaproxy/issues/1689
Can you please tell me where I can add more docs and the JSON schema for e.g.
[OWASP ZAP FuzzAPI.pdf](https://github.com/zaproxy/zap-extensions/files/3869143/OWASP.ZAP.FuzzAPI.pdf)
If there are any issues regarding the code or any changes are required, I am happy to work further.
Thanks

Commit:
Initialized API class

davy320 committed:
Fuzzer dialog opening

davy320 committed:
Fuzzer dialog opening

davy320 committed:
Fuzzer dialog opening

Commit:
Added input parameters to simpleHTTPFuzzer

davy320 committed:
Added Json Http Message Input functionality

davy320 committed:
Merge branch 'master' of https://github.com/davy320/zap-extensions

 Conflicts:
	addOns/fuzz/src/main/java/org/zaproxy/zap/extension/fuzz/FuzzAPI.java

davy320 committed:
Merging done, with the SimpleHttpFuzzer

davy320 committed:
Merging done, with the SimpleHttpFuzzer

davy320 committed:
Fixed minor issues.

Commit:
Added view to display progress of fuzzer

davy320 committed:
Fixed minor issues.

davy320 committed:
Merge branch 'master' of https://github.com/davy320/zap-extensions

 Conflicts:
	addOns/fuzz/src/main/java/org/zaproxy/zap/extension/fuzz/FuzzAPI.java

davy320 committed:
merged, fuzz status

davy320 committed:
Deleted reload_Extenstion not required

davy320 committed:
Implemented Json Payload input

davy320 committed:
input from json as a fuzz location totally works :)

davy320 committed:
input from json as a fuzz location totally works :)

davy320 committed:
processors added

davy320 committed:
pretty sure everything is working just fine now

davy320 committed:
Added multiple features check ids and scripts

davy320 committed:
Fixed minor issues, or redundant structures

Commit:
Added support for api client libraries generation

davy320 committed:
Added some javadocs, comments, and added common strings to Message.properties

davy320 committed:
fixed the last bug of multiple payload json fuzzer not working

davy320 committed:
Added some javadocs, comments, and added common strings to Message.properties

davy320 committed:
FuzzOptions work well now probably.

davy320 committed:
Removed some redundant code in FuzzAPI

davy320 committed:
Merge branch 'master' of https://github.com/zaproxy/zap-extensions

davy320 committed:
Deleted build_zap.sh

Unnecessary file, build specific.

davy320 committed:
Updated the change log.

davy320 committed:
Added "GET_ALL_SENT_MESSSAGES" to the API calls

davy320 committed:
Added logger, and proper json responses for invalid inputs.

davy320 committed:
Merge branch 'master' of https://github.com/davy320/zap-extensions

 Conflicts:
	addOns/fuzz/CHANGELOG.md
	addOns/fuzz/src/main/java/org/zaproxy/zap/extension/fuzz/FuzzAPI.java
	addOns/fuzz/src/main/resources/org/zaproxy/zap/extension/fuzz/resources/Messages.properties

davy320 committed:
Merge branch 'master' of https://github.com/zaproxy/zap-extensions

davy320 committed:
Added help documentation for fuzzAPI

davy320 committed:
Merge branch 'master' of https://github.com/zaproxy/zap-extensions

kingthorin commented on addOns/fuzz/src/main/java/org/zaproxy/zap/extension/fuzz/FuzzAPI.java:
```diff
@@ -0,0 +1,842 @@
+        return PREFIX;
+    }
+
+    /** resets the api "only" fuzzerOptionsToDefault */
```
This comment seems un-necessary.

Though arguably having jdoc is good, if you leave it please format it properly and use title caps appropriately.

kingthorin commented on addOns/fuzz/src/main/java/org/zaproxy/zap/extension/fuzz/FuzzAPI.java:
```diff
@@ -0,0 +1,842 @@
+                TableHistory tableHistory = Model.getSingleton().getDb().getTableHistory();
+                RecordHistory recordHistory =
+                        getRecordHistory(tableHistory, getParam(params, PARAM_MESSAGE_ID, -1));
+                assert (recordHistory.getHttpMessage() != null);
```
I believe this will return. Should throw api exception?

kingthorin commented on addOns/fuzz/src/main/java/org/zaproxy/zap/extension/fuzz/FuzzAPI.java:
```diff
@@ -0,0 +1,842 @@
+                        getParam(params, PARAM_STRATEGY, null),
+                        getParam(params, PARAM_RETRIES, -1),
+                        getParam(params, PARAM_DELAY, -1),
+                        getParam(params, PARAM_THREADS, -1));
```
Will threads -1 work?

Will all these work defaulted to -1?

Edit: I guess this is addressed at Line 403?

kingthorin commented on addOns/fuzz/src/main/java/org/zaproxy/zap/extension/fuzz/FuzzAPI.java:
```diff
@@ -0,0 +1,842 @@
+            case ACTION_SET_HTTP_FUZZ_OPTIONS:
+                setHttpFuzzerOptions(
+                        getParam(params, PARAM_MAX_ERRORS_ALLOWED, -1),
+                        getParam(params, PARAM_STRATEGY, null),
```
Default null or `""` empty string?

kingthorin commented on addOns/fuzz/src/main/java/org/zaproxy/zap/extension/fuzz/FuzzAPI.java:
```diff
@@ -0,0 +1,842 @@
+            case ACTION_PAUSE_SCAN:
+                httpFuzzer = getFuzzer(params);
+                httpFuzzer.pauseScan();
+                return ApiResponseElement.OK;
```
Does this need to check that it was successful?

kingthorin commented on addOns/fuzz/src/main/java/org/zaproxy/zap/extension/fuzz/FuzzAPI.java:
```diff
@@ -0,0 +1,842 @@
+            case ACTION_RESUME_SCAN:
+                httpFuzzer = getFuzzer(params);
+                httpFuzzer.resumeScan();
+                return ApiResponseElement.OK;
```
Does this need to check that it was successful?

kingthorin commented on addOns/fuzz/src/main/java/org/zaproxy/zap/extension/fuzz/FuzzAPI.java:
```diff
@@ -0,0 +1,842 @@
+        try {
+            allLines = Files.readAllLines(Paths.get(payloadPath));
+        } catch (IOException e) {
+            e.printStackTrace();
```
Should log properly and throw api exception?

kingthorin commented on addOns/fuzz/src/main/java/org/zaproxy/zap/extension/fuzz/FuzzAPI.java:
```diff
@@ -0,0 +1,842 @@
+                return HttpMessage.class;
+            }
+            // There is no need for these but can be fixed
+            // All of the functions below no need probably
```
?

kingthorin commented on addOns/fuzz/src/main/java/org/zaproxy/zap/extension/fuzz/FuzzAPI.java:
```diff
@@ -0,0 +1,842 @@
+     *
+     * @see HttpFuzzerHandler /* Used to create the fuzzer
+     */
+    /* Didn't want to change fuzzerHandler to public so reusing it */
```
These should be combined and formatted properly.

Is there a reason it shouldn't be public? (Or protected)

davy320 replied:
Yeah there is no reason for it to be not public. 
Just wanted to confirm.

kingthorin replied:
@thc202 / @psiinon thoughts on making it public?

thc202 replied:
It's fine making it public, noting that's not part of the API.

kingthorin commented on addOns/fuzz/src/main/java/org/zaproxy/zap/extension/fuzz/FuzzAPI.java:
```diff
@@ -0,0 +1,842 @@
+     * Method copied from HttpFuzzerHandler
+     *
+     * @see HttpFuzzerHandler Used to shrink the name of the fuzzer
+     */
```
Is there a reason it shouldn't be public? (Or protected)

kingthorin commented on addOns/fuzz/src/main/java/org/zaproxy/zap/extension/fuzz/FuzzAPI.java:
```diff
@@ -0,0 +1,842 @@
+    }
+
+    // TODO try and implement this Start using this
+    // Testing required
```
?

kingthorin commented on addOns/fuzz/src/main/java/org/zaproxy/zap/extension/fuzz/FuzzAPI.java:
```diff
@@ -0,0 +1,842 @@
+                findCharactersInsideString(messageObject.get("requestHeader").toString());
+        List<Integer> bodyFuzzLocations =
+                findCharactersInsideString(messageObject.get("requestBody").toString());
+        // Remove the SECTION_SIGN character not needed anymore
```
?

kingthorin commented on addOns/fuzz/src/main/java/org/zaproxy/zap/extension/fuzz/FuzzAPI.java:
```diff
@@ -0,0 +1,842 @@
+    }
+
+    /**
+     * Tries to create a net.sf.jsonobject from the local file path If found create the object and
```
`...path if found...` (lower i on if)

kingthorin commented on addOns/fuzz/src/main/java/org/zaproxy/zap/extension/fuzz/FuzzAPI.java:
```diff
@@ -0,0 +1,842 @@
+     * @param jsonPath this is the local path in the environment which contains a valid net.sf json
+     *     object
+     * @return net sf Jsonobject
+     */
```
Should be consistent on `net.sf.jsonobject`

kingthorin replied:
Maybe even `{@code net.sf.jsonobject}`

kingthorin commented on addOns/fuzz/src/main/java/org/zaproxy/zap/extension/fuzz/FuzzAPI.java:
```diff
@@ -0,0 +1,842 @@
+        try {
+            is = new FileInputStream(initialFile);
+        } catch (FileNotFoundException e) {
+            e.printStackTrace();
```
Should log properly

kingthorin commented on addOns/fuzz/src/main/java/org/zaproxy/zap/extension/fuzz/FuzzAPI.java:
```diff
@@ -0,0 +1,842 @@
+        try {
+            jsonTxt = IOUtils.toString(is);
+        } catch (IOException e) {
+            e.printStackTrace();
```
Should log properly

kingthorin commented on addOns/fuzz/src/main/java/org/zaproxy/zap/extension/fuzz/FuzzAPI.java:
```diff
@@ -0,0 +1,842 @@
+     * This message is not being used right now maybe implementation for a future release //TODO
+     *
+     * @param messageObject
+     * @return
```
Should be completed properly

kingthorin commented on addOns/fuzz/src/main/java/org/zaproxy/zap/extension/fuzz/FuzzAPI.java:
```diff
@@ -0,0 +1,842 @@
+                    new HttpMessage(
+                            requestHeader, requestBodyObject, responseHeader, responseBodyObject);
+        } catch (HttpMalformedHeaderException e) {
+            e.printStackTrace();
```
Should log properly

kingthorin commented on addOns/fuzz/src/main/java/org/zaproxy/zap/extension/fuzz/FuzzAPI.java:
```diff
@@ -0,0 +1,842 @@
+    }
+
+    // Check if the signum character was escaped
+    // TODO implement this
```
?

kingthorin commented on addOns/fuzz/src/main/java/org/zaproxy/zap/extension/fuzz/FuzzAPI.java:
```diff
@@ -0,0 +1,842 @@
+
+    /**
+     * This methods reads the Json file which contains the fuzzLocations and the different payloads
+     * types and locations It follows a particular schema which can be obtained from the repository
```
Is there a missing period or should "it" be lower cased?

kingthorin commented on addOns/fuzz/src/main/java/org/zaproxy/zap/extension/fuzz/FuzzAPI.java:
```diff
@@ -0,0 +1,842 @@
+     * @link <a href="https://github.com/zaproxy/zap-api-docs">API DOCS<a/>
+     * @param fuzzLocationsObject This is the json object from the file that was uploaded
+     * @return List of all the fuzzLocations and their pre-generated payloads
+     * @version 1.0
```
Should be removed

kingthorin commented on addOns/fuzz/src/main/javahelp/org/zaproxy/zap/extension/fuzz/resources/help/contents/api.html:
```diff
@@ -0,0 +1,173 @@
+<H1>Fuzzing API</H1>
+<p>
+    The API has been designed to be used by people with a range of security experience, for
+    example “simpleHttpFuzzer” is a basic fuzzer fuzzes one “Fuzz location”, extracting payloads
```
`...is a basic fuzzer which fuzzes one..`

kingthorin commented on addOns/fuzz/src/main/javahelp/org/zaproxy/zap/extension/fuzz/resources/help/contents/api.html:
```diff
@@ -0,0 +1,173 @@
+</p>
+<h3>Message Processors</h3>
+<p>
+    The default processor enabled by default for all requests is “Request Content-Length Updater”
```
Should this be "default" or "mandatory" as I recall the earlier code the user isn't able to unset it.

davy320 replied:
True, the user is not able to unset it. Changing it to The requisite processor.

kingthorin commented on addOns/fuzz/src/main/javahelp/org/zaproxy/zap/extension/fuzz/resources/help/contents/api.html:
```diff
@@ -0,0 +1,173 @@
+The parameter fuzzLocation is the start and end number of the character which is replaced
+by the payload.<br>
+For example:
+<p style="font-family:'Courier 10 Pitch',cursive">POST /cgi-bin/process.cgi HTTP/1.1
```
This should probably use `<TT>` or `<code>`... @thc202 what needs to be used here to parse properly for the github wiki pages?

kingthorin replied:
Maybe `<pre>` instead of leading the lines with `<br>` (which seems really strange)?

thc202 replied:
Yeah, better to use `<code>` and `<pre>` instead of changing font and adding all the `br` tags.

kingthorin commented on addOns/fuzz/src/main/javahelp/org/zaproxy/zap/extension/fuzz/resources/help/contents/api.html:
```diff
@@ -0,0 +1,173 @@
+Note: options changed after the fuzzer has been initialised will remain same for that fuzzer.
+<h3>Simple Http Fuzzer</h3>
+<p>The simpleHttpFuzzer fuzzes one fuzz location, with inputs</p>
+<table style="width:100%" border="1px solid black" border-collapse="collapse">
```
Isn't 100% default?

davy320 replied:
It is not 100% default, but I still removed it, as it is not required.

kingthorin commented on addOns/fuzz/src/main/javahelp/org/zaproxy/zap/extension/fuzz/resources/help/contents/api.html:
```diff
@@ -0,0 +1,173 @@
+    fuzzer run.
+</p>
+
+<table style="width:100%" border="1px solid black" border-collapse="collapse">
```
Isn't 100% default?

kingthorin commented on addOns/fuzz/src/main/javahelp/org/zaproxy/zap/extension/fuzz/resources/help/contents/api.html:
```diff
@@ -0,0 +1,173 @@
+        <br class="SCXW166475849 BCX2">
+      </span><span class="TextRun SCXW166475849 BCX2" data-contrast="none" style="color: rgb(0, 0, 0); font-size: 10pt; font-family: Consolas, Consolas_MSFontService, monospace; line-height: 16px;" lang="EN-AU"><span class="NormalTextRun SCXW166475849 BCX2" style="background-color: inherit;">}</span></span><span class="EOP SCXW166475849 BCX2" data-ccp-props="{&quot;201341983&quot;:0,&quot;3355597 ...
+    </div>
+</div>
```
There must be a cleaner way to do this.

kingthorin commented on addOns/fuzz/src/main/javahelp/org/zaproxy/zap/extension/fuzz/resources/help/contents/api.html:
```diff
@@ -0,0 +1,173 @@
+</div>
+
+<p>This way the fuzzer accepts multiple fuzz locations and multiple payload types.</p>
+<p>More of these examplesand the JSON schema can be found at zap-api-docs:&nbsp;</p>
```
`examplesand `

kingthorin commented on addOns/fuzz/src/main/javahelp/org/zaproxy/zap/extension/fuzz/resources/help/contents/api.html:
```diff
@@ -0,0 +1,173 @@
+<p>This way the fuzzer accepts multiple fuzz locations and multiple payload types.</p>
+<p>More of these examplesand the JSON schema can be found at zap-api-docs:&nbsp;</p>
+<p><a href="https://github.com/zaproxy/zap-api-docs/fuzz">https://github.com/zaproxy/zap-api-docs/fuzz</a>/</p>
+<p>The JSON schema is version controlled and the json should be updated according to the version of the fuzzer being used.</p>
```
If this is true should the parsing to moved to a utility class with appropriate methods (instead of being included monolithically in FuzzAPI).

davy320 replied:
Just removed the version controlling, not necessary, works well.

kingthorin commented on addOns/fuzz/src/main/resources/org/zaproxy/zap/extension/fuzz/resources/Messages.properties:
```diff
@@ -387,8 +387,6 @@ fuzz.httpfuzzer.description = Allows to fuzz HTTP messages.
 fuzz.httpfuzzer.messagetype = HTTP
 fuzz.httpfuzzer.fuzzerNamePrefix = HTTP - {0}
 
-fuzz.httpfuzzer.searcher.name = HTTP Fuzzer Results
```
Why was this removed? It's used in ExtesionHttpFuzzer.java

davy320 replied:
Yes, there were 2 copies of the same key/value pair.

kingthorin replied:
Confirmed, thanks!
![image](https://user-images.githubusercontent.com/7570458/69349716-3edaeb80-0c46-11ea-9dd8-1603bed8b3a7.png)

kingthorin commented on addOns/fuzz/src/main/resources/org/zaproxy/zap/extension/fuzz/resources/Messages.properties:
```diff
@@ -484,3 +482,13 @@ fuzz.httpfuzzer.processor.tagcreator.validation.messageboxtitle = Please verify
 fuzz.httpfuzzer.processor.tagcreator.validation.regexatleastonegroup = The entered regular expression needs at least one group.
+
+fuzz.api.action.multiplePayloadFuzzer = Upload the json file containing fuzz, payload and processors locations, the schema can be obtained from zap-api-docs\n.
+fuzz.api.action.simpleHttpFuzzer = Simple Http Fuzzer fuzzes one location, with inputs messageId {int}, requestLocation {body or header}, fuzzLocation, character at {int start: int end}, payload {Path of the file containing payloads}. Example: messageID: 7, requestLocation: "body", fuzzLocation: "4:8", payload: "/home/users/payloads.txt"
```
`A simple Http Fuzzer which fuzzes one location...`

kingthorin commented on addOns/fuzz/src/main/resources/org/zaproxy/zap/extension/fuzz/resources/Messages.properties:
```diff
@@ -484,3 +482,13 @@ fuzz.httpfuzzer.processor.tagcreator.validation.messageboxtitle = Please verify
+fuzz.api.response.failure.getfuzzer = Failed to get fuzzer
+fuzz.api.jsoninput.location = location
+fuzz.api.jsoninput.fuzzlocations = fuzzLocations
+fuzz.api.jsoninput.start = start
```
No end?

kingthorin commented on addOns/fuzz/src/main/resources/org/zaproxy/zap/extension/fuzz/resources/Messages.properties:
```diff
@@ -484,3 +482,13 @@ fuzz.httpfuzzer.processor.tagcreator.validation.messageboxtitle = Please verify
 fuzz.httpfuzzer.processor.tagcreator.validation.regexsyntaxerror = The entered regular expression has a syntax error:\r\n{0}
 fuzz.httpfuzzer.processor.tagcreator.validation.regexatleastonegroup = The entered regular expression needs at least one group.
+
+fuzz.api.action.multiplePayloadFuzzer = Upload the json file containing fuzz, payload and processors locations, the schema can be obtained from zap-api-docs\n.
```
JSON in full caps?

Should the line feed really be before the period?

kingthorin commented on addOns/fuzz/src/main/javahelp/org/zaproxy/zap/extension/fuzz/resources/help/contents/api.html:
```diff
@@ -0,0 +1,173 @@
+<p>In this HTTP POST request, the top and bottom are header and body respectively.
+    <br><br>To fuzz the value of licenseID in the body the value of string needs to be replaced.
+</p>
+    <img src="images/simple_http_fuzzer_example.png" alt="Simple Http Fuzzer Example" width="300">
```
Maintaining this image might be a headache down the road. Perhaps quote the same example as the messages.properties file?
`Example: messageID: 7, requestLocation: "body", fuzzLocation: "4:8", payload: "/home/users/payloads.txt"`

kingthorin replied:
@thc202 @psiinon thoughts?

thc202 replied:
Agreed, better not add the image, I'd also question the need for the simple endpoint.

davy320 replied:
Removed the image, not required.

kingthorin commented on addOns/fuzz/src/main/javahelp/org/zaproxy/zap/extension/fuzz/resources/help/contents/api.html:
```diff
@@ -0,0 +1,173 @@
+</p>
+<h3>Multiple Payload Fuzzer</h3>
+<p>The multiple payload fuzzer as per name fuzzes multiple fuzz Locations with multiple payload
+    types. It takes input as the message Id and a certain Json file with a schematic to make the
```
JSON in caps?

(Probably applies elsewhere as well....in written form I believe acronyms should use full caps)

kingthorin commented:
Obviously the [LGTM issues](https://github.com/zaproxy/zap-extensions/pull/2222#issuecomment-555988817) should also be addressed :+1:

davy320 commented on addOns/fuzz/src/main/java/org/zaproxy/zap/extension/fuzz/FuzzAPI.java:
```diff
@@ -0,0 +1,842 @@
+                if (fuzzer == null) {
+                    return new ApiResponseElement(
+                            Constant.messages.getString("fuzz.api.response.failure.getfuzzer"));
+                }
```
Done thanks, and also it was necessary to check fuzzerId for Actions, so combined it. :)

davy320 committed:
Fixed TODO logs, removed unwanted implementations, docs, and rendundancy fixed.

kingthorin commented on addOns/fuzz/src/main/resources/org/zaproxy/zap/extension/fuzz/resources/Messages.properties:
```diff
@@ -484,3 +480,7 @@ fuzz.httpfuzzer.processor.tagcreator.validation.messageboxtitle = Please verify
 fuzz.httpfuzzer.processor.tagcreator.validation.regexatleastonegroup = The entered regular expression needs at least one group.
+
+fuzz.api.action.multiplePayloadFuzzer = Upload the JSON file containing fuzz, payload and processors locations, the schema can be obtained from zap-api-docs.
+fuzz.api.action.simpleHttpFuzzer = A Simple Http Fuzzer fuzzes one location, with inputs messageId {int}, requestLocation {body or header}, fuzzLocation, character at {int start: int end}, payload {Path of the file containing payloads}. Example: messageID: 7, requestLocation: "body", fuzzLocation: "4:8", payload: "/home/users/payloads.txt"
```
`A simple Http Fuzzer which fuzzes one location...` add "which"..."Fuzzer fuzzes" just looks funny, not grammatically wrong, just looks out of place.

kingthorin reviewed (commented):
Thanks.

davy320 commented:
> Thanks.

So I fixed the LGTM issues, but I am unsure why they are not showing up in the results that they have been fixed?

kingthorin commented:
Don't worry about it, it's a weird thing the bot does (or well doesn't do) sometimes. The updated analysis is here > https://lgtm.com/projects/g/zaproxy/zap-extensions/rev/pr-1f2844ec82fa17b68cf86e73903944194d67d753 and I agree it's showing as no issues now. (I'll delete the out dated LGTM comment above.)

davy320 committed:
Fixed TODO logs, removed unwanted implementations, docs, and rendundancy fixed.

rscottbailey committed:
Merge remote-tracking branch 'zaproxy/master' into reconcile

rscottbailey committed:
getId() -> getIndex()

Enough to fix gradlew errors after merging with mainline, anyway

rscottbailey committed:
Bump version

...so we can distinguish this modified version from the stock addon

rscottbailey commented:
I have some feedback on this request (which I hope could be merged in some fashion in the near future, as my docker image is beginning to look like Frankenstein's monster...)

* I have made edits to bring the PR source current with master as of last week. (This involved fixing the merge conflict with CHANGELOG.md noted above, and changing some method names to track the "Id" -> "Index" transition that evidently occurred in mainline somewhere after November.
* These edits were submitted to davy320 via PR on 15 May 2020, but I have not seen any response; you can find them directly in rscottbailey/zap-extensions branch reconcile
* I have successfully exercised most (but not all) of these API endpoints from the web-based API browser and also a locally-hacked version of python-owasp-zap-v2.4. "most" means "I haven't bothered to exercise ALL of the endpoints yet" -- no failures detected except as noted in the next bullet.
* One bizarre oddity I still am studying: Starting /zap/zap.sh with "-daemon" causes all fuzz API endpoints to return "internal_error"; I am guessing there is still something somewhere in fuzz that is tied to UI initialization and getting skipped in headless mode.

I am very motivated to help move this PR forward, but I am unsure what will be most effective. Please me me know.

Thanks!

thc202 commented:
Is there any exception in the log when the internal error happens?
Which endpoints are you using?

There are some changes that need to be done before merging, e.g.:
 - This should probably moved to the HTTP implementation since it's just about HTTP;
 - The JSON schema needs to be reviewed to make sure it can handle (or accommodate) all the features that are supported in the UI.

rscottbailey commented:
When executing with a UI, this call succeeds:

```python
                fuzz_id = zap.fuzz.multiple_payload_fuzzer(
                    messageid=message["id"],
                    jsonfuzzlocationsfilepath=temp_name,
                )
```

When executing in headless (`-daemon`) mode, the same call fails. I find in zap.log:

```
2020-05-18 18:25:25,989 [ZAP-ProxyThread-467] INFO  FuzzAPI - Starting fuzzer
2020-05-18 18:25:25,994 [ZAP-ProxyThread-467] INFO  FuzzAPI - Running multiple payload fuzzer.
2020-05-18 18:25:25,994 [ZAP-ProxyThread-467] INFO  FuzzAPI - Fuzzer options used are: 5 100 1000 0 DEPTH_FIRST
2020-05-18 18:25:25,994 [ZAP-ProxyThread-467] INFO  FuzzAPI - Number of fuzzLocations: 1
2020-05-18 18:25:25,994 [ZAP-ProxyThread-467] INFO  HttpFuzzer - Fuzzer started...
2020-05-18 18:25:25,995 [ZAP-ProxyThread-467] ERROR API - Exception while handling API request:
java.lang.NullPointerException
        at org.zaproxy.zap.extension.fuzz.ExtensionFuzz.runFuzzer(ExtensionFuzz.java:680)
        at org.zaproxy.zap.extension.fuzz.FuzzAPI.handleApiAction(FuzzAPI.java:310)
        at org.zaproxy.zap.extension.api.API.handleApiRequest(API.java:506)
        at org.parosproxy.paros.core.proxy.ProxyThread.processHttp(ProxyThread.java:499)
        at org.parosproxy.paros.core.proxy.ProxyThread.run(ProxyThread.java:335)
        at java.base/java.lang.Thread.run(Thread.java:834)
2020-05-18 18:25:26,000 [ZAP-FuzzerTerminationNotifier-433] INFO  HttpFuzzer - Fuzzer completed.
```

With respect to @thc202's asks:

* This should probably moved to the HTTP implementation since it's just about HTTP;

I am unclear on to what "it" refers. The target being fuzzed? The communication channel to the daemon? Isn't *EVERYTHING* about zaproxy basically http[s]?

* The JSON schema needs to be reviewed to make sure it can handle (or accommodate) all the features that are supported in the UI.

I believe better minds than mine conducted this review back in November 2019... Technically, the answer for multiple_payload_fuzzer is "yes" as the UI sets a pretty low functionality bar. I am developing a python wrapper (guess what my application uses) that adds some useful functionality such as "fuzz messages matching this regex" (vs "choose the message to fuzz...") and "determine pass/fail based on response code" (vs "you must manually inspect the results...")

I am uncertain how best (or even if) those would be added to the actual zaproxy API, so I am looking first to get #2222 in so I have something to work from as I continue development.

Cheers,
Scott

thc202 commented:
> I find in zap.log:

Thanks for the stacktrace, yeah, that's because it's still accessing the UI (as it was originally expected to have one).

> I am unclear on to what "it" refers.

Referring to the API being added.

> Isn't EVERYTHING about zaproxy basically http[s]?

No, e.g. WebSocket.

> I believe better minds than mine conducted this review back in November 2019... 

I don't see any comments about the JSON schema here (there was some comments in the issue but nothing in depth e.g. how are the payload processors defined for a specific payload generator, or all; message processors, are they identified by an ID, which one? classname?).

> Technically, the answer for multiple_payload_fuzzer is "yes" as the UI sets a pretty low functionality bar.

With that endpoint how do you specify a message processor? What about other message locations (e.g. JWT being added in other add-on)?

I guess we could say that the current version is good to start, but it would be great to try accommodate what it's currently supported in the UI.

davy320 committed:
Merge pull request #1 from rscottbailey/reconcile

Merge with upstream/master

davy320 commented:
> 
> 
> When executing with a UI, this call succeeds:
> 
> ```python
>                 fuzz_id = zap.fuzz.multiple_payload_fuzzer(
>                     messageid=message["id"],
>                     jsonfuzzlocationsfilepath=temp_name,
>                 )
> ```
> 
> When executing in headless (`-daemon`) mode, the same call fails. I find in zap.log:
> 
> ```
> 2020-05-18 18:25:25,989 [ZAP-ProxyThread-467] INFO  FuzzAPI - Starting fuzzer
> 2020-05-18 18:25:25,994 [ZAP-ProxyThread-467] INFO  FuzzAPI - Running multiple payload fuzzer.
> 2020-05-18 18:25:25,994 [ZAP-ProxyThread-467] INFO  FuzzAPI - Fuzzer options used are: 5 100 1000 0 DEPTH_FIRST
> 2020-05-18 18:25:25,994 [ZAP-ProxyThread-467] INFO  FuzzAPI - Number of fuzzLocations: 1
> 2020-05-18 18:25:25,994 [ZAP-ProxyThread-467] INFO  HttpFuzzer - Fuzzer started...
> 2020-05-18 18:25:25,995 [ZAP-ProxyThread-467] ERROR API - Exception while handling API request:
> java.lang.NullPointerException
>         at org.zaproxy.zap.extension.fuzz.ExtensionFuzz.runFuzzer(ExtensionFuzz.java:680)
>         at org.zaproxy.zap.extension.fuzz.FuzzAPI.handleApiAction(FuzzAPI.java:310)
>         at org.zaproxy.zap.extension.api.API.handleApiRequest(API.java:506)
>         at org.parosproxy.paros.core.proxy.ProxyThread.processHttp(ProxyThread.java:499)
>         at org.parosproxy.paros.core.proxy.ProxyThread.run(ProxyThread.java:335)
>         at java.base/java.lang.Thread.run(Thread.java:834)
> 2020-05-18 18:25:26,000 [ZAP-FuzzerTerminationNotifier-433] INFO  HttpFuzzer - Fuzzer completed.
> ```
> 
> With respect to @thc202's asks:
> 
>     * This should probably moved to the HTTP implementation since it's just about HTTP;
> 
> 
> I am unclear on to what "it" refers. The target being fuzzed? The communication channel to the daemon? Isn't _EVERYTHING_ about zaproxy basically http[s]?
> 
>     * The JSON schema needs to be reviewed to make sure it can handle (or accommodate) all the features that are supported in the UI.
> 
> 
> I believe better minds than mine conducted this review back in November 2019... Technically, the answer for multiple_payload_fuzzer is "yes" as the UI sets a pretty low functionality bar. I am developing a python wrapper (guess what my application uses) that adds some useful functionality such as "fuzz messages matching this regex" (vs "choose the message to fuzz...") and "determine pass/fail based on response code" (vs "you must manually inspect the results...")
> 
> I am uncertain how best (or even if) those would be added to the actual zaproxy API, so I am looking first to get #2222 in so I have something to work from as I continue development.
> 
> Cheers,
> Scott

Hi Scott, please have a look at the status of the build after I merged your code and if you need some help, let me know.

rscottbailey commented:
@davy320 I have PR #2 against your repo with a few transformations I missed because I initially was only trying to build the one fuzz module. As noted in PR, some openapi unit tests are failing, but I don't understand how that would be related to any of these changes (or if it is something weird about my development environment). Let me know if this works any better.

On a slightly different note, in-house testing of API-driven fuzzing is going well. (Successfully spidering target site, running active scan, then programmatically selecting and fuzzing earlier traffic and retrieving results.) This is unfortunately slow, because if I just blast away blindly with fuzz requests, zap wedges after about 16 or so of them -- so I am paranoid and wait for one to complete before proceeding to the next. This is difficult because fuzzerProgress() reports in terms of messages sent... (Is there in fact any difference between it and getMessagesSentCount?) Is there a better way to see a scan has finished than to poll fuzzerProgress() and infer the scan is inactive when the returned count stops changing? Especially when referencing one of the predefined payload lists, my code has no way to know how many messages are going to be sent.

I would look more and ask left if I were more fluent in java LOL -- I'm attempting to avoid falling into false conclusions due to misunderstanding what I am reading.

rscottbailey committed:
Revert "getId() -> getIndex()"

This reverts commit 44b0495bcf08eabd4a78559196001c602ba52161.

rscottbailey committed:
Revert "Bump version"

This reverts commit c5f3e8e80d74083afeb5caa31e5016d5169ebf80.

rscottbailey committed:
Merge branch 'master' of github.com:zaproxy/zap-extensions

Incorporate #2397 to fix test failures

rscottbailey committed:
Merge branch 'master' into rsbredo

Pull upstream unit test fixes (#2397)

rscottbailey committed:
Merge pull request #1 from rscottbailey/rsbredo

Back out unnecessary edits

rscottbailey committed:
Add support for fuzzer message processor management (#2)

Add fuzz API support for:

* Adding a script to list of fuzzer message processors (at an arbitrary point)
* Removing a fuzzer message processor (by name)
* Resetting the fuzzer message processor list to a known default
* Displaying the list of fuzz message processors

The API functions for running fuzzers have been updated to use this list of message processors instead of an internal hard-coded list.

Default behavior (if no explicit message processor changes are made) is basically the same, except that the payload reflection processor has been added because I saw it was in the default list of processors in the UI.

Like options, the processor list is managed internally and used only when creating new http fuzzers; list changes do not affect existing fuzzers.

Signed-off-by: Scott Bailey <scott.bailey@godaddy.com>

rscottbailey committed:
Add support for fuzzer message processor management (#2)

Add fuzz API support for:

* Adding a script to list of fuzzer message processors (at an arbitrary point)
* Removing a fuzzer message processor (by name)
* Resetting the fuzzer message processor list to a known default
* Displaying the list of fuzz message processors

The API functions for running fuzzers have been updated to use this list of message processors instead of an internal hard-coded list.

Default behavior (if no explicit message processor changes are made) is basically the same, except that the payload reflection processor has been added because I saw it was in the default list of processors in the UI.

Like options, the processor list is managed internally and used only when creating new http fuzzers; list changes do not affect existing fuzzers.

rscottbailey commented:
I have just sent @davy320 a PR that extends the API to cover fuzzer message processor manipulation.

This is working in my environment and hopefully is not so horrid that merging will be problematic. It comes closer to addressing @thc202's request for feature parity with the UI. (This extension only supports script-based message processors... I could barely get this working without complicating it by supporting arbitrary subtypes! :O)

davy320 committed:
Merge pull request #3 from rscottbailey/master

Extend Fuzz API to manage http fuzz processors by rscottbailey.

davy320 committed:
Merge branch 'master' into master

rscottbailey committed:
Merge remote-tracking branch 'upstream/master'

davy320 committed:
Merge pull request #4 from rscottbailey/master

API extensions, revisited

rscottbailey commented:
Any tips on working through the DCO issues? I spent an hour or two attempting to add the signoff to (just) my commits, thought I succeeded on the one that actually did something, but the analyzer doesn't seem to be placated. It seems rather against the spirit of the DCO to "shotgun-signoff" a grab-bag of 50+ commits I know nothing about.

Mostly I'm just trying to figure what could be done to push this PR closer to the finish line.

kingthorin commented:
To address the DCO requirement you'll need to sign-off the commit(s):

- https://github.com/zaproxy/zaproxy/blob/develop/CONTRIBUTING.md#developer-certificate-of-origin
- https://git-scm.com/docs/git-commit#Documentation/git-commit.txt---signoff

thc202 commented:
Older commits would have to be signed-off too to pass the check. DCO is not a blocker though.

> Mostly I'm just trying to figure what could be done to push this PR closer to the finish line.

Discuss some of the concerns above:
https://github.com/zaproxy/zap-extensions/pull/2222#issuecomment-630379235
https://github.com/zaproxy/zap-extensions/pull/2222#issuecomment-630419824

>  @thc202's request for feature parity with the UI

To be clear, it would be great to have full parity but I'm not requesting (nor expecting) that, what I'm asking is that the current JSON schema at least allows to be improved upon without having to rewrite/break the JSON schema (i.e. the questions asked in a previous comment would lead to breakages or be harder to improve).

rscottbailey commented:
> There are some changes that need to be done before merging, e.g.:
> 
> * This should probably moved to the HTTP implementation since it's just about HTTP;

Unless @davy320 has knowledge and time, I'm going to need a sketch of what this means, since I feel lucky to make my incremental changes work at all LOL...

> * The JSON schema needs to be reviewed to make sure it can handle (or accommodate) all the features that are supported in the UI.

I presume we are discussing multiplePayloadFuzzer's JSON payload, since that is most analogous to the existing UI. My (admittedly biased) review:

Starting with the `Fuzz Locations` tab (after having selected a message):

* The UI requires you to select a portion of the chosen message that will be fuzzed. It is possible to fuzz multiple locations. In the API, the equivalent action is to populate `fuzzLocations`. This is a list, and therefore can accommodate multiple locations, like the UI. Each location is identified by `requestLocation`, `start` and `end`, providing precision equivalent to the UI.
* The UI then allows you to add (or remove or modify) payloads. The API uses the `payloads` list in each entry to specify one or more payloads, each of which consists of a `type` and additional type-specific members. I have only used string and file fuzzers, so I don't know that all of the types supported by the UI are supported by the API.
* I don't really understand the function of the payload-specific "Processors..." dialog, but I don't see anything in the JSON that addresses this area.
* In the UI, you click the `Start Fuzzer` button; Making the `multiplePayloadFuzzers` call starts a fuzzer, and there are endpoints to check the status of a fuzzer and stop an in-progress fuzzer.
* With respect to extensibility, the JSON schema includes both `schema` and `version` elements, so I think it would be straightforward to do version bumps when the schema is extended to add more functionality.

The following stuff doesn't really have any relationship to the JSON payload, but for completeness...

The UI `Options` tab has an analogous API `setHttpFuzzerOptions` endpoint, which supports all of the displayed options except for the `Follow Redirects` checkbox. There's a `resetHttpFuzzOptionsToDefault` endpoint which does the same thing as the UI's `Reset` button.

The UI `Message Processors` tab is paralleled by the contribution I made, almost. You can add script processors to arbitrary locations in the message processor list, delete processors (including the default non-script processors), and reset the list to a default setting (which matches the UI default list). What you *cannot* do with the UI is (a) reverse the order of the `Request Content-Length Updater` and `Payload Reflection Detector` -- but this makes no practical difference IMHO -- or add a "Tag Creator" processor.

Hopefully this is useful.

tarkatronic commented:
Hi @davy320 I just wanted to check in on here -- are you still committed to getting this PR fixed up and merged? I'm working with @rscottbailey now on our in-house solution which depends on the functionality in this PR, and it would be wonderful to see this brought across the finish line!

Please let me know what we can do to ensure this gets the attention it needs. Thank you!

davy320 commented:
> 
> 
> Hi @davy320 I just wanted to check in on here -- are you still committed to getting this PR fixed up and merged? I'm working with @rscottbailey now on our in-house solution which depends on the functionality in this PR, and it would be wonderful to see this brought across the finish line!
> 
> Please let me know what we can do to ensure this gets the attention it needs. Thank you!

Hi, 

@tarkatronic, I am looking forward to working on this, It's still a lot of work to get this production ready.

The first time I created this I got all the CICD tests passed but before that I would like to work more on making it work on base64, general blob, and other types on inputs that the ZAP fuzzer accepts but it's not included in the plugin I wrote years back.

Let's make it more generalised in a ZAP way. Whatever fuzz options are available in ZAP should also work with the API. 

Cheers,
Dennis

Jeeppler commented:
@davy320 are you still working on this issue?

psiinon commented:
Thank you for this PR, and sorry it has taken us so long to address it.
We have decided to close it, but if you are interested in revisiting it, or in contributing to ZAP in any way then please get in touch.
